### PR TITLE
Add multi-block transform functionality.

### PR DESCRIPTION
**Pull request by jorgefilipecosta**
## Description
This PR's implements multi-block transform functionality. Allowing the user to transform a selection of multiple blocks into other block types. It changes blocks api. When specifying the transform the developer can set isMultiBlock to true to set that the transform is able to handle more than one block.
Multiblock transforms were already added to transform from multiple paragraphs into a list and multiple images into a gallery.

## Testing
Create multiple paragraphs, select them, go to the transform area and transform them into a list.
Add multiple images select them and transform them into a gallery.

## Screenshots:
nov-06-2017 22-55-50
nov-06-2017 22-59-24

### CODE REVIEW

File: editor/block-switcher/index.js [hunk 1]
```diff
@@ -113,14 +118,20 @@ function BlockSwitcher( { block, onTransform } ) {
...
 
 export default connect(
-	( state, ownProps ) => ( {
-		block: getBlock( state, ownProps.uid ),
-	} ),
-	( dispatch, ownProps ) => ( {
-		onTransform( block, name ) {
+	( state, ownProps ) => {
+		const isMultiBlock = ! ownProps.uid && getSelectedBlockCount( state ) > 1;
+		const blocks = isMultiBlock ? getMultiSelectedBlocks( state ) :
```
Comment (youknowriad) [inline editor/block-switcher/index.js]: It feels a bit weird to retrive the blocks from the state when there's a multiselection and retrieve them from the `uid` when passed. Can we update the component to always use a `uids` prop To make this consistent?

File: blocks/api/test/factory.js [hunk 2]
```diff
@@ -144,7 +144,7 @@ describe( 'block factory', () => {
 				value: 'ribs',
 			} );
 
-			const transformedBlocks = switchToBlockType( block, 'core/updated-text-block' );
+			const transformedBlocks = switchToBlockType( [ block ], 'core/updated-text-block' );
```
Comment (youknowriad) [inline blocks/api/test/factory.js]: To avoid all these changes, we could have used lodash's `castArray` in the `switchToBlockType` function

**Review by youknowriad — COMMENTED** [review]
This is really awesome, I like it.

- Now that we have this, I think we should update the `list -> paragraph` single transformation to create multiple paragraphs, one for each list item. And for consistency, I'd also update the `paragraph -> list` transformation to always create a list with one item (and avoid splitting on brs) cc @jasmussen 

- Images to gallery is great too, but same here, we should now add a `gallery => images` transform.

Overall, the code looks good, I like the approach. Really great work here

**Comment by jasmussen** [comment]
> Now that we have this, I think we should update the list -> paragraph single transformation to create multiple paragraphs, one for each list item. And for consistency, I'd also update the paragraph -> list transformation to always create a list with one item (and avoid splitting on brs)

This sounds good to me. Given the recent improvements in cross-block selection, and the hope that we can make multi-select transformations like selecting three paragraphs and transforming those into a list, this makes a lot of sense.

transform multiple

I think @mtias noted an opportunity around lists as well, that if we were to make each list item into its own block, we might be able to do some cool list sorting things as part of that.

**Comment by youknowriad** [comment]
This probably closes #539

File: editor/block-switcher/index.js [hunk 3]
```diff
@@ -24,19 +24,25 @@ import { getBlock } from '../selectors';
...
-	const blockType = getBlockType( block.name );
+	if ( isMultiBlock && ! every( blocks, ( block ) => ( block.name === sourceBlockName ) ) ) {
+		return null;
+	}
+
+	const blockType = getBlockType( sourceBlockName );
 	const blocksToBeTransformedFrom = reduce( getBlockTypes(), ( memo, type ) => {
 		const transformFrom = get( type, 'transforms.from', [] );
-		const transformation = find( transformFrom, t => t.type === 'block' && t.blocks.indexOf( block.name ) !== -1 );
+		const transformation = find( transformFrom, t => t.type === 'block' && t.blocks.indexOf( sourceBlockName ) !== -1 && ( ! isMultiBlock || t.isMultiBlock ) );
```
Comment (youknowriad) [inline editor/block-switcher/index.js]: Minor: The line is a bit too long

File: editor/header/header-toolbar/index.js [hunk 4]
```diff
@@ -35,6 +36,7 @@ function HeaderToolbar( { hasUndo, hasRedo, hasFixedToolbar, undo, redo } ) {
 				label={ __( 'Redo' ) }
 				disabled={ ! hasRedo }
 				onClick={ redo } />
+			{ isMultiBlockSelection && <BlockSwitcher key="switcher" uids={ selectedBlockUids } /> }
```
Comment (youknowriad) [inline editor/header/header-toolbar/index.js]: From a design perspective, this is always showing up in the Header even if I choose to show the block toolbar next to the block, I think we might want to show the switcher in the first selected block's toolbar when this option is toggled. (You can toggle this option in the ellipsis menu at the top right of the editor) cc @jasmussen

Comment (jorgefilipecosta) [reply editor/header/header-toolbar/index.js]: Regarding this point, I agree, adding the transform drop-down to the tools of the first block may be an option if the config preference is set. But I think adding this logic would increase the scope of this PR and make it harder to review. If it is ok for you I would prefer to add this logic in a future PR.

**Review by youknowriad — APPROVED** [review]
Awesome PR

**Comment by jasmussen** [comment]
I think the work done here is very nice, and yes it does indeed close #539. 

The current mockups put the transformation into the ellipsis menu, like Riad suggests. But an argument can be made that it's okay to show this in the toolbar. If it's significantly easier to do so, we may do that as a first step before we put the transformation inside the ellipsis menu, then we can test this as an interface first, and move on to the ellipsis if we find that is more contextual and sufficiently discoverable.

Let's discuss the perspectives first, though, and @karmatosed and @mtias feel free to chime in:

- We'd want to show list making chiefly on paragraph blocks. That works fine ✅
- We'd want you to be able to also make blockquotes when selecting _virtually_ any block. I say "virtually" because there might be added complexity for images that are wide or fullwide, and until we have proper nesting we'd probably want to limit this to just a few block types.
- In the future it would be nice to be able to select 5 images in sequence, and transform those into a gallery.
- Code or Preformatted would also be nice options to allow on blocks outside of just paragraphs.
- Perhaps it would be nice to create a plugin API, so plugins could show their own buttons next to some multi selections?

None of the above have to be addressed in _this_ PR, but it's context for deciding whether we want to move the transformations into an ellipsis now, or later. Like I said, if it's a ton of extra work, it's probably okay to try and go with this now, but at a minimum we'd want to fix this extra space that appears:

extra space

**Comment by jasmussen** [comment]
A quick next enhancement could be to explode the contents of the switcher dropdown into direct buttons for heading, quote, list, pre and verse, as opposed to keeping them in a dropdown.

**Comment by jasmussen** [comment]
Also agree: awesome PR! Thank you!

**Comment by youknowriad** [comment]
> In the future it would be nice to be able to select 5 images in sequence, and transform those into a gallery.

This already works :)

**Comment by youknowriad** [comment]
>  whether we want to move the transformations into an ellipsis now

Actually, I was not talking about moving those to the ellipsis for now. I was just saying we should be consistent with the position of the block switcher of one selected block: It shows in the header's toolbar if the block toolbar is "fixed" and shows in the contextual toolbar if it's not.

--

That said, I think moving all these switchers to the block's ellipsis menu should be better but I also think this is a separate task :)

**Comment by jasmussen** [comment]
If Riad approves so do I!

**Comment by jorgefilipecosta** [comment]
Thank you for your inputs it was merged :) Now the work will continue to address the other points.

**Comment by mtias** [comment]
Great work here. I agree with moving this to the ellipsis menu for the selected group of blocks. I actually missed the transformation in the header.

**Pull request merged**